File: packages/praxrr-app/src/lib/server/utils/logger/logger.ts
/**
 * Logger with console and file output
 * Supports configurable settings and daily rotation
 * Can run independently with provided config or use system defaults
 */

import { config } from '$config';
import { colors } from './colors.ts';
import { logSettings } from './settings.ts';
import type { LogEntry, LoggerConfig, LogLevel, LogOptions } from './types.ts';

class Logger {
  private config: Required<LoggerConfig>;

  constructor(config?: LoggerConfig) {
    // Use provided config or sensible defaults
    this.config = {
      logsDir: config?.logsDir ?? '/tmp/logs',
      enabled: config?.enabled ?? true,
      fileLogging: config?.fileLogging ?? true,
      consoleLogging: config?.consoleLogging ?? true,
      minLevel: config?.minLevel ?? 'INFO',
    };
  }

  private formatTimestamp(): string {
    const timestamp = new Date().toISOString();
    return `${colors.grey}${timestamp}${colors.reset}`;
  }

  private formatLevel(level: string, color: string): string {
    return `${color}${level.padEnd(5)}${colors.reset}`;
  }

  private formatSource(source?: string): string {
    if (!source) return '';
    return `${colors.grey}[${source}]${colors.reset}`;
  }

  private formatMeta(meta?: unknown): string {
    if (!meta) return '';
    return `${colors.grey}${JSON.stringify(meta)}${colors.reset}`;
  }

  /**
   * Get log file path with daily rotation (YYYY-MM-DD.log)
   */
  private getLogFilePath(): string {
    const date = new Date().toISOString().split('T')[0]; // YYYY-MM-DD
    return `${this.config.logsDir}/${date}.log`;
  }

  /**
   * Check if logging is enabled
   */
  private isEnabled(): boolean {
    const currentSettings = logSettings.get();
    return currentSettings.enabled === 1;
  }

  /**
   * Check if file logging is enabled
   */
  private isFileLoggingEnabled(): boolean {
    const currentSettings = logSettings.get();
    return currentSettings.file_logging === 1;
  }

  /**
   * Check if console logging is enabled
   */
  private isConsoleLoggingEnabled(): boolean {
    const currentSettings = logSettings.get();
    return currentSettings.console_logging === 1;
  }

  /**
   * Check if a log level should be logged based on minimum level
   */
  private shouldLog(level: LogLevel): boolean {
    if (!this.isEnabled()) {
      return false;
    }

    // Get fresh settings from database instead of using cached config
    const currentSettings = logSettings.get();
    const currentMinLevel = currentSettings.min_level;

    const levels: LogLevel[] = ['DEBUG', 'INFO', 'WARN', 'ERROR'];
    const minIndex = levels.indexOf(currentMinLevel);
    const levelIndex = levels.indexOf(level);

    return levelIndex >= minIndex;
  }

  private async log(level: LogLevel, color: string, message: string, options?: LogOptions): Promise<void> {
    // Check if this log level should be logged
    if (!this.shouldLog(level)) {
      return;
    }

    const timestamp = new Date().toISOString();

    // Console output (colored)
    if (this.isConsoleLoggingEnabled()) {
      const consoleParts = [
        this.formatTimestamp(),
        this.formatLevel(level, color),
        message,
        options?.source ? this.formatSource(options.source) : '',
        options?.meta ? this.formatMeta(options.meta) : '',
      ].filter(Boolean);

      console.log(consoleParts.join(' | '));
    }

    // File output (JSON)
    if (this.isFileLoggingEnabled()) {
      const logEntry: LogEntry = {
        timestamp,
        level,
        message,
        ...(options?.source ? { source: options.source } : {}),
        ...(options?.meta ? { meta: options.meta } : {}),
      };

      try {
        const filePath = this.getLogFilePath();

        // Ensure logs directory exists
        try {
          await Deno.mkdir(this.config.logsDir, { recursive: true });
        } catch {
          // Directory might already exist
        }

        // Write to log file
        await Deno.writeTextFile(filePath, JSON.stringify(logEntry) + '\n', {
          append: true,
        });
      } catch (error) {
        // If file write fails, at least we have console output
        console.error('Failed to write to log file:', error);
      }
    }
  }

  async debug(message: string, options?: LogOptions): Promise<void> {
    await this.log('DEBUG', colors.cyan, message, options);
  }

  async info(message: string, options?: LogOptions): Promise<void> {
    await this.log('INFO', colors.green, message, options);
  }

  async warn(message: string, options?: LogOptions): Promise<void> {
    await this.log('WARN', colors.yellow, message, options);
  }

  async error(message: string, options?: LogOptions): Promise<void> {
    await this.log('ERROR', colors.red, message, options);
  }

  async errorWithTrace(message: string, error?: Error, options?: LogOptions): Promise<void> {
    await this.log('ERROR', colors.red, message, options);

    // Print stack trace to console
    if (error?.stack && this.isConsoleLoggingEnabled()) {
      console.log(`${colors.grey}${error.stack}${colors.reset}`);
    }

    // Write stack trace to file
    if (error?.stack && this.isFileLoggingEnabled()) {
      const traceEntry: LogEntry = {
        timestamp: new Date().toISOString(),
        level: 'ERROR',
        message: 'Stack trace',
        meta: { stack: error.stack },
      };

      try {
        const filePath = this.getLogFilePath();
        await Deno.writeTextFile(filePath, JSON.stringify(traceEntry) + '\n', {
          append: true,
        });
      } catch (writeError) {
        console.error('Failed to write stack trace to log file:', writeError);
      }
    }
  }
}

// Export Logger class for creating custom instances (for testing)
export { Logger };

/**
 * Default logger singleton for production use
 * Uses system config and database settings
 *
 * For testing, create a standalone Logger instance:
 * @example
 * import { Logger } from './logger.ts';
 * const testLogger = new Logger({ logsDir: '/tmp/test-logs', minLevel: 'DEBUG' });
 * await testLogger.info('test message');
 */
const settings = logSettings.get();
export const logger = new Logger({
  logsDir: config.paths.logs,
  enabled: settings.enabled === 1,
  fileLogging: settings.file_logging === 1,
  consoleLogging: settings.console_logging === 1,
  minLevel: settings.min_level,
});

File: packages/praxrr-parser/Logging/Logger.cs
using System.Text.Json;

namespace Parser.Logging;

/// <summary>
/// Logger with console and file output
/// Supports configurable settings and daily rotation
/// </summary>
public class Logger
{
    private readonly LoggerConfig _config;
    private readonly LogSettingsManager? _settings;
    private static readonly object _fileLock = new();

    public Logger(LoggerConfig? config = null, LogSettingsManager? settings = null)
    {
        _settings = settings;
        _config = config ?? _settings?.Get() ?? new LoggerConfig();
    }

    private string FormatTimestamp()
    {
        var timestamp = DateTime.UtcNow.ToString("o");
        return $"{Colors.Grey}{timestamp}{Colors.Reset}";
    }

    private string FormatLevel(LogLevel level)
    {
        var color = level switch
        {
            LogLevel.DEBUG => Colors.Cyan,
            LogLevel.INFO => Colors.Green,
            LogLevel.WARN => Colors.Yellow,
            LogLevel.ERROR => Colors.Red,
            _ => Colors.Reset
        };
        return $"{color}{level.ToString().PadRight(5)}{Colors.Reset}";
    }

    private string FormatSource(string? source)
    {
        if (string.IsNullOrEmpty(source)) return "";
        return $"{Colors.Grey}[{source}]{Colors.Reset}";
    }

    private string FormatMeta(object? meta)
    {
        if (meta == null) return "";
        return $"{Colors.Grey}{JsonSerializer.Serialize(meta)}{Colors.Reset}";
    }

    /// <summary>
    /// Get log file path with daily rotation (YYYY-MM-DD.log)
    /// </summary>
    private string GetLogFilePath()
    {
        var date = DateTime.UtcNow.ToString("yyyy-MM-dd");
        return Path.Combine(_config.LogsDir, $"{date}.log");
    }

    private bool IsEnabled()
        => _settings?.IsEnabled() ?? _config.Enabled;

    private bool IsFileLoggingEnabled()
        => _settings?.IsFileLoggingEnabled() ?? _config.FileLogging;

    private bool IsConsoleLoggingEnabled()
        => _settings?.IsConsoleLoggingEnabled() ?? _config.ConsoleLogging;

    private bool ShouldLog(LogLevel level)
    {
        if (!IsEnabled()) return false;
        var minLevel = _settings?.GetMinLevel() ?? _config.MinLevel;
        return level >= minLevel;
    }

    private void Log(LogLevel level, string message, LogOptions? options = null)
    {
        if (!ShouldLog(level)) return;

        var timestamp = DateTime.UtcNow.ToString("o");

        // Console output (colored)
        if (IsConsoleLoggingEnabled())
        {
            var parts = new List<string>
            {
                FormatTimestamp(),
                FormatLevel(level),
                message
            };

            if (!string.IsNullOrEmpty(options?.Source))
                parts.Add(FormatSource(options.Source));

            if (options?.Meta != null)
                parts.Add(FormatMeta(options.Meta));

            Console.WriteLine(string.Join(" | ", parts));
        }

        // File output (JSON)
        if (IsFileLoggingEnabled())
        {
            var logEntry = new LogEntry
            {
                Timestamp = timestamp,
                Level = level.ToString(),
                Message = message,
                Source = options?.Source,
                Meta = options?.Meta
            };

            try
            {
                // Ensure logs directory exists
                Directory.CreateDirectory(_config.LogsDir);

                var filePath = GetLogFilePath();
                var json = JsonSerializer.Serialize(logEntry) + Environment.NewLine;

                // Thread-safe file write
                lock (_fileLock)
                {
                    File.AppendAllText(filePath, json);
                }
            }
            catch (Exception ex)
            {
                // If file write fails, at least we have console output
                Console.Error.WriteLine($"Failed to write to log file: {ex.Message}");
            }
        }
    }

    public void Debug(string message, LogOptions? options = null)
        => Log(LogLevel.DEBUG, message, options);

    public void Debug(string message, string source)
        => Log(LogLevel.DEBUG, message, new LogOptions { Source = source });

    public void Info(string message, LogOptions? options = null)
        => Log(LogLevel.INFO, message, options);

    public void Info(string message, string source)
        => Log(LogLevel.INFO, message, new LogOptions { Source = source });

    public void Warn(string message, LogOptions? options = null)
        => Log(LogLevel.WARN, message, options);

    public void Warn(string message, string source)
        => Log(LogLevel.WARN, message, new LogOptions { Source = source });

    public void Error(string message, LogOptions? options = null)
        => Log(LogLevel.ERROR, message, options);

    public void Error(string message, string source)
        => Log(LogLevel.ERROR, message, new LogOptions { Source = source });

    public void Error(string message, Exception ex, LogOptions? options = null)
    {
        Log(LogLevel.ERROR, message, options);

        // Print stack trace to console
        if (ex.StackTrace != null && IsConsoleLoggingEnabled())
        {
            Console.WriteLine($"{Colors.Grey}{ex.StackTrace}{Colors.Reset}");
        }

        // Write stack trace to file
        if (ex.StackTrace != null && IsFileLoggingEnabled())
        {
            var traceEntry = new LogEntry
            {
                Timestamp = DateTime.UtcNow.ToString("o"),
                Level = "ERROR",
                Message = "Stack trace",
                Meta = new { stack = ex.StackTrace }
            };

            try
            {
                var filePath = GetLogFilePath();
                var json = JsonSerializer.Serialize(traceEntry) + Environment.NewLine;

                lock (_fileLock)
                {
                    File.AppendAllText(filePath, json);
                }
            }
            catch (Exception writeEx)
            {
                Console.Error.WriteLine($"Failed to write stack trace to log file: {writeEx.Message}");
            }
        }
    }
}

/// <summary>
/// Global logger singleton for production use
/// </summary>
public static class Log
{
    private static Logger? _instance;

    public static void Initialize(LoggerConfig? config = null, LogSettingsManager? settings = null)
    {
        _instance = new Logger(config, settings);
    }

    public static Logger Instance =>
        _instance ?? throw new InvalidOperationException("Logger not initialized. Call Initialize() first.");

    // Convenience methods that delegate to the singleton
    public static void Debug(string message, LogOptions? options = null)
        => Instance.Debug(message, options);

    public static void Debug(string message, string source)
        => Instance.Debug(message, source);

    public static void Info(string message, LogOptions? options = null)
        => Instance.Info(message, options);

    public static void Info(string message, string source)
        => Instance.Info(message, source);

    public static void Warn(string message, LogOptions? options = null)
        => Instance.Warn(message, options);

    public static void Warn(string message, string source)
        => Instance.Warn(message, source);

    public static void Error(string message, LogOptions? options = null)
        => Instance.Error(message, options);

    public static void Error(string message, string source)
        => Instance.Error(message, source);

    public static void Error(string message, Exception ex, LogOptions? options = null)
        => Instance.Error(message, ex, options);
}
